skbio/diversity/beta/_unifrac.py

```diff
@@ -0,0 +1,170 @@
+
+    Parameters
+    ----------
+    u_counts, v_counts: list, np.array
+        Vectors of counts of OTUs for two samples. Must be equal length.
+    otu_ids: list, np.array
+        Vector of OTU ids corresponding to tip names in ``tree``. Must be the
+        same length as ``u_counts`` and ``v_counts``.
+    tree: skbio.TreeNode
+        Tree relating the OTUs in otu_ids. The set of tip names in the tree can
+        be a superset of ``otu_ids``, but not a subset.
+    normalized: boolean, optional
+        If ``True``, apply branch length normalization. Resulting distances
+        will then be in the range ``[0, 1]``.
+
+    Returns
+    -------
+    float
+        The weighted unifrac distance between the two samples.
+
+    Raises
+    ------
+    ValueError
+        If ``u_counts``, ``v_counts``, and ``otu_ids`` are not all equal in
+        length.
+    MissingNodeError
+        If an OTU id is provided that does not correspond to a tip in the
+        tree.
+
+    See Also
+    --------
+    unweighted_unifrac
+
+    Notes
+    -----
+    Unweighted unifrac was originally described in [1]_, which includes a
+    discussion of unweighted (qualitative) versus weighted (quantitiative)
+    diversity metrics.
+
+    References
+    ----------
+    .. [1] Lozupone, C. A., Hamady, M., Kelley, S. T. & Knight, R. Quantitative
+       and qualitative beta diversity measures lead to different insights into
+       factors that structure microbial communities. Appl. Environ. Microbiol.
+       73, 1576–1585 (2007).
+
+    """
+    u_observed_otus = {o: c for o, c in zip(otu_ids, u_counts) if c >= 1}
+    u_total_count = sum(u_counts)
+    v_observed_otus = {o: c for o, c in zip(otu_ids, v_counts) if c >= 1}
+    v_total_count = sum(v_counts)
+    if u_total_count == 0 and v_total_count == 0:
+        # boundary case where both communities have no members
+        return 0.0
+    u_observed_nodes = tree.observed_node_counts(u_observed_otus)
+    v_observed_nodes = tree.observed_node_counts(v_observed_otus)
+    weighted_unifrac = 0
+    D = 0
+    for o in set(u_observed_nodes) | set(v_observed_nodes):
+        # handle the case of o.length is None
+        b = o.length or 0
+        u_branch_weight = _sample_branch_weight(
+            u_observed_nodes[o], u_total_count)
+        v_branch_weight = _sample_branch_weight(
+            v_observed_nodes[o], v_total_count)
+        branch_weight = abs(u_branch_weight - v_branch_weight)
+        weighted_unifrac += b * branch_weight
+        if o.is_tip() and normalized:
+            d = o.accumulate_to_ancestor(tree.root())
+            normed_weight = u_branch_weight + v_branch_weight
+            D += (d * normed_weight)
+    if normalized:
+        return weighted_unifrac / D
+    else:
+        return weighted_unifrac
+
+
+def _sample_branch_weight(observed_nodes, total_count):
+    if total_count == 0:
+        return 0
```

Comment:
Correct. One note on weighted unifrac: we may want to investigate other ways of doing the difference in branch weight, e.g. squared differences. There are some other variants of unifrac out there e.g. VAW that we may also want to look into, if only to benchmark.

> On Aug 7, 2015, at 9:41 AM, Greg Caporaso <notifications@github.com> wrote:
> 
> In skbio/diversity/beta/_unifrac.py <https://github.com/biocore/scikit-bio/pull/1027#discussion_r36536638>:
> 
> > +            v_observed_nodes[o], v_total_count)
> > +        branch_weight = abs(u_branch_weight - v_branch_weight)
> > +        weighted_unifrac += b * branch_weight
> > +        if o.is_tip() and normalized:
> > +            d = o.accumulate_to_ancestor(tree.root())
> > +            normed_weight = u_branch_weight + v_branch_weight
> > +            D += (d * normed_weight)
> > +    if normalized:
> > +        return weighted_unifrac / D
> > +    else:
> > +        return weighted_unifrac
> > +
> > +
> > +def _sample_branch_weight(observed_nodes, total_count):
> > +    if total_count == 0:
> > +        return 0
> It wouldn't just be PD (I think), b/c of weighting.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/biocore/scikit-bio/pull/1027/files#r36536638>.
>